packages/nuxt/src/core/runtime/nitro/renderer.ts

```diff
@@ -498,15 +498,17 @@ function joinTags (tags: string[]) {
 }
 
 function joinAttrs (chunks: string[]) {
-  return chunks.join(' ')
+  return chunks.join(' ').trim()
 }
 
 function renderHTMLDocument (html: NuxtRenderHTMLContext) {
-  return `<!DOCTYPE html>
-<html ${joinAttrs(html.htmlAttrs)}>
-<head>${joinTags(html.head)}</head>
-<body ${joinAttrs(html.bodyAttrs)}>${joinTags(html.bodyPrepend)}${joinTags(html.body)}${joinTags(html.bodyAppend)}</body>
-</html>`
+  return [
+    '<!DOCTYPE html>',
```

Comment:
What about instead using `+` rather than Array.join? Ive not benchmarked but think it might be a bit more performant.

Reply:
Thanks for the reminder, I've updated the implementation.

I tried to test three implementation methods through the `perf.link` website. The result is as follows: 

👉 [renderHTMLDocument - perf.link](https://perf.link/#eyJpZCI6IjVwbnNqZzFuNWwzIiwidGl0bGUiOiJyZW5kZXJIVE1MRG9jdW1lbnQiLCJiZWZvcmUiOiJmdW5jdGlvbiBqb2luVGFncyAodGFncykge1xuICByZXR1cm4gdGFncy5qb2luKCcnKVxufVxuXG5mdW5jdGlvbiBqb2luQXR0cnMgKGNodW5rcykge1xuICByZXR1cm4gY2h1bmtzLmpvaW4oJyAnKVxufVxuXG5mdW5jdGlvbiByZW5kZXJIVE1MRG9jdW1lbnQxIChodG1sKSB7XG4gIHJldHVybiBgPCFET0NUWVBFIGh0bWw%2BXG48aHRtbCAke2pvaW5BdHRycyhodG1sLmh0bWxBdHRycyl9PlxuPGhlYWQ%2BJHtqb2luVGFncyhodG1sLmhlYWQpfTwvaGVhZD5cbjxib2R5ICR7am9pbkF0dHJzKGh0bWwuYm9keUF0dHJzKX0%2BJHtqb2luVGFncyhodG1sLmJvZHlQcmVwZW5kKX0ke2pvaW5UYWdzKGh0bWwuYm9keSl9JHtqb2luVGFncyhodG1sLmJvZHlBcHBlbmQpfTwvYm9keT5cbjwvaHRtbD5gXG59XG5cbmZ1bmN0aW9uIHJlbmRlckhUTUxEb2N1bWVudDIgKGh0bWwpIHtcbiAgcmV0dXJuICc8IURPQ1RZUEUgaHRtbD4nXG4gICAgKyBgPGh0bWwgJHtqb2luQXR0cnMoaHRtbC5odG1sQXR0cnMpfWAudHJpbSgpICsgJz4nXG4gICAgKyBgPGhlYWQ%2BJHtqb2luVGFncyhodG1sLmhlYWQpfTwvaGVhZD5gXG4gICAgKyBgPGJvZHkgJHtqb2luQXR0cnMoaHRtbC5ib2R5QXR0cnMpfWAudHJpbSgpICsgYD4ke2pvaW5UYWdzKGh0bWwuYm9keVByZXBlbmQpfSR7am9pblRhZ3MoaHRtbC5ib2R5KX0ke2pvaW5UYWdzKGh0bWwuYm9keUFwcGVuZCl9PC9ib2R5PmBcbiAgICArICc8L2h0bWw%2BJ1xufVxuXG5mdW5jdGlvbiByZW5kZXJIVE1MRG9jdW1lbnQzIChodG1sKSB7XG4gIHJldHVybiBbXG4gICAgJzwhRE9DVFlQRSBodG1sPicsXG4gICAgYDxodG1sICR7am9pbkF0dHJzKGh0bWwuaHRtbEF0dHJzKX1gLnRyaW0oKSArICc%2BJyxcbiAgICBgPGhlYWQ%2BJHtqb2luVGFncyhodG1sLmhlYWQpfTwvaGVhZD5gLFxuICAgIGA8Ym9keSAke2pvaW5BdHRycyhodG1sLmJvZHlBdHRycyl9YC50cmltKCkgKyBgPiR7am9pblRhZ3MoaHRtbC5ib2R5UHJlcGVuZCl9JHtqb2luVGFncyhodG1sLmJvZHkpfSR7am9pblRhZ3MoaHRtbC5ib2R5QXBwZW5kKX08L2JvZHk%2BYCxcbiAgICAnPC9odG1sPidcbiAgXS5qb2luKCcnKVxufVxuXG5jb25zdCBodG1sID0ge1xuICAgIGlzbGFuZDogZmFsc2UsXG4gICAgaHRtbEF0dHJzOiBbJyddLFxuICAgIGhlYWQ6IFtcbiAgICAgICc8bWV0YSBjaGFyc2V0PVwidXRmLThcIj5cXG48bWV0YSBuYW1lPVwidmlld3BvcnRcIiBjb250ZW50PVwid2lkdGg9ZGV2aWNlLXdpZHRoLCBpbml0aWFsLXNjYWxlPTFcIj5cXG48bGluayByZWw9XCJwcmVsb2FkXCIgYXM9XCJmZXRjaFwiIGNyb3Nzb3JpZ2luPVwiYW5vbnltb3VzXCIgaHJlZj1cIi9fcGF5bG9hZC5qc29uXCI%2BXFxuPGxpbmsgcmVsPVwibW9kdWxlcHJlbG9hZFwiIGFzPVwic2NyaXB0XCIgY3Jvc3NvcmlnaW4gaHJlZj1cIi9fbnV4dC9lbnRyeS41dURoQnNIYS5qc1wiPlxcbjxsaW5rIHJlbD1cInByZWZldGNoXCIgYXM9XCJzY3JpcHRcIiBjcm9zc29yaWdpbiBocmVmPVwiL19udXh0L2Vycm9yLTQwNC5GdU1xMEYwVi5qc1wiPlxcbjxsaW5rIHJlbD1cInByZWZldGNoXCIgYXM9XCJzY3JpcHRcIiBjcm9zc29yaWdpbiBocmVmPVwiL19udXh0L3Z1ZS5mMzZhY2QxZi5XRGpUNDlHbS5qc1wiPlxcbjxsaW5rIHJlbD1cInByZWZldGNoXCIgYXM9XCJzY3JpcHRcIiBjcm9zc29yaWdpbiBocmVmPVwiL19udXh0L2Vycm9yLTUwMC5Tb2FXeFhHSi5qc1wiPlxcbjxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi9fbnV4dC9lbnRyeS41dURoQnNIYS5qc1wiIGNyb3Nzb3JpZ2luPjwvc2NyaXB0PidcbiAgICAgIF0sXG4gICAgYm9keUF0dHJzOiBbJyddLFxuICAgIGJvZHlQcmVwZW5kOiBbXSxcbiAgICBib2R5OiBbXG4gICAgICAnPGRpdiBpZD1cIl9fbnV4dFwiPjxkaXY%2BIE51eHQgMyBQbGF5Z3JvdW5kIDwvZGl2PjwvZGl2PidcbiAgICBdLFxuICAgIGJvZHlBcHBlbmQ6IFtcbiAgICAgICc8c2NyaXB0IHR5cGU9XCJhcHBsaWNhdGlvbi9qc29uXCIgaWQ9XCJfX05VWFRfREFUQV9fXCIgZGF0YS1zc3I9XCJ0cnVlXCIgZGF0YS1zcmM9XCIvX3BheWxvYWQuanNvblwiPlt7XCJzdGF0ZVwiOjEsXCJvbmNlXCI6MyxcIl9lcnJvcnNcIjo1LFwic2VydmVyUmVuZGVyZWRcIjo3LFwicGF0aFwiOjgsXCJwcmVyZW5kZXJlZEF0XCI6OX0sW1wiUmVhY3RpdmVcIiwyXSx7fSxbXCJSZWFjdGl2ZVwiLDRdLFtcIlNldFwiXSxbXCJSZWFjdGl2ZVwiLDZdLHt9LHRydWUsXCIvXCIsMTcwMzQ4Mzk4OTQ1NF08L3NjcmlwdD5cXG48c2NyaXB0PndpbmRvdy5fX05VWFRfXz17fTt3aW5kb3cuX19OVVhUX18uY29uZmlnPXtwdWJsaWM6e30sYXBwOntiYXNlVVJMOlwiL1wiLGJ1aWxkQXNzZXRzRGlyOlwiL19udXh0L1wiLGNkblVSTDpcIlwifX08L3NjcmlwdD4nXG4gICAgXVxuICB9IiwidGVzdHMiOlt7Im5hbWUiOiJvcmlnaW5hbCIsImNvZGUiOiJyZW5kZXJIVE1MRG9jdW1lbnQxKGh0bWwpIiwicnVucyI6Wzc4MDAwMCw4MTUwMDAsMTA0MDAwLDExNzAwMCwxNzk1MDAwLDYyMzAwMCwxMjkwMDAsMTY0MDAwLDU5NzAwMCw2MzkwMDAsMTQxNzAwMCwyOTYzMDAwLDc2ODAwMCwyMDY1MDAwLDEzNDgwMDAsMTUzMjAwMCwyMTM5MDAwLDMyMDAwLDE2NzEwMDAsMzU1MDAwLDIzNzcwMDAsMjA1OTAwMCwyNzc2MDAwLDMyNzIwMDAsODQ0MDAwLDc0NjAwMCwxNzQwMDAwLDE4NjcwMDAsMjA2MjAwMCwyMjI4MDAwLDEyMDAwMCwxMDAwLDE0ODUwMDAsMTI5MDAwLDgzMDAwLDIzNDYwMDAsMTQ3NzAwMCwxOTUwMDAwLDIwNzYwMDAsMjUyMTAwMCwxNDg0MDAwLDIzOTMwMDAsNzg1MDAwLDMyNjAwMDAsNDg5MDAwLDE5MTAwMDAsMTg0MDAwMCwxNjk2MDAwLDMyNTYwMDAsNDE0ODAwMCwxMjkwMDAsMTQ0NzAwMCw4MTYwMDAsMjIyNDAwMCw1MzMwMDAsMTgyMDAwMCwxODMyMDAwLDI3MzIwMDAsMjA2NDAwMCwyMzgzMDAwLDMzNjAwMDAsMjcxNjAwMCw1OTcwMDAsOTQyMDAwLDM1ODYwMDAsNDAwMCw4MjAwMCwyMDg4MDAwLDI4MDkwMDAsNjc3MDAwLDY2NDAwMCwyMDczMDAwLDE5ODcwMDAsMTQwNTAwMCwxOTUwMDAsMTEyMTAwMCw0MTQ1MDAwLDE4MDEwMDAsMTAwMCwxMjQ0MDAwLDMzMjgwMDAsMzE4MzAwMCwxMjkwMDAsMjE4NzAwMCwyMDQ4MDAwLDIxNDIwMDAsMzAxODAwMCwyMDM5MDAwLDExNzMwMDAsMTg4MjAwMCwyMjA3MDAwLDE2ODgwMDAsMzQ2MDAwLDI1OTAwMCwyNjg0MDAwLDM4NzgwMDAsMjcxMDAwLDI0MzYwMDAsMTkxMDAwMCwxODI0MDAwXSwib3BzIjoxNTk1ODIwfSx7Im5hbWUiOiJVc2UgYCtgIiwiY29kZSI6InJlbmRlckhUTUxEb2N1bWVudDIoaHRtbCkiLCJydW5zIjpbMjA3MDAwLDI3MDAwMCwxMDAwLDcyMDAwLDk2OTAwMCwxMjkwMDAsMTMwMDAsNzYwMDAsMzM0MDAwLDQwMzAwMCwxMjkwMDAsMjExNTAwMCw5NTkwMDAsMTQ2MTAwMCw4MDgwMDAsMTQyNjAwMCw3MzIwMDAsMjgwMDAsMTA5NDAwMCwzMTAwMDAsMTk4NjAwMCwyMTAzMDAwLDE4MzUwMDAsMjM3OTAwMCw3MzEwMDAsNTgyMDAwLDExODcwMDAsMTMxODAwMCwxMDg3MDAwLDE3MzkwMDAsOTgwMDAsMjE3NzAwMCwxMTEzMDAwLDEyOTAwMCwyMjY2MDAwLDE2NTgwMDAsOTQ4MDAwLDExMTYwMDAsMTQ0MTAwMCwxNjM1MDAwLDIxMDIwMDAsMTc2OTAwMCw0NzEwMDAsMjQ1OTAwMCwzODQwMDAsMTM0OTAwMCwxOTM5MDAwLDEyMDkwMDAsMjgxMjAwMCwyODk1MDAwLDEyOTAwMCwxMTEyMDAwLDU5NjAwMCwxMjg5MDAwLDQxNzAwMCwxMDAwLDE1NTYwMDAsMTkyNDAwMCw0NTMwMDAsMjIzNDAwMCwyNDY2MDAwLDIxMzIwMDAsMTM3MzAwMCw2OTQwMDAsMjcwMDAwMCwxNzg2MDAwLDcwMDAwLDE0NjUwMDAsMjE3ODAwMCwzOTUwMDAsMzIyMDAwLDEzNjYwMDAsMTUzNDAwMCw4OTAwMDAsMTY0MDAwLDcyNTAwMCwyODk1MDAwLDEzOTEwMDAsMTAwMCw3MjcwMDAsMjM0MTAwMCwyNDcyMDAwLDk3MDAwLDE2MzYwMDAsMTQ3MjAwMCwxNDMwMDAwLDI4OTUwMDAsMjAzMDAwMCw4MzIwMDAsMTUzOTAwMCwxNjY4MDAwLDE1NDQwMDAsMzA5MDAwLDI0MzAwMCwyMjg2MDAwLDI4ODAwMDAsMjQ5MDAwLDE4NzgwMDAsMTMyMDAwMCwxMjA1MDAwXSwib3BzIjoxMjE3NjQwfSx7Im5hbWUiOiJVc2UgYEFycmF5LmpvaW5gIiwiY29kZSI6InJlbmRlckhUTUxEb2N1bWVudDMoaHRtbCkiLCJydW5zIjpbMjU1MDAwLDE2MzAwMCwzMzgwMDAsNjEwMDAsNDQ3MDAwLDUwMDAwMCw2MzYwMDAsNTkzMDAwLDE4MjAwMCwxMzkwMDAsNjc5MDAwLDY3OTAwMCw1NjIwMDAsNjc5MDAwLDE4MjAwMCw1NjMwMDAsNTI1MDAwLDE1MDAwLDYzMDAwMCwxOTIwMDAsNTEyMDAwLDY3OTAwMCw2NzkwMDAsMTE2NjAwMCw0ODAwMDAsMzIyMDAwLDUxNDAwMCwzNjAwMDAsNjc5MDAwLDUwMzAwMCw3MjQwMDAsNjc5MDAwLDU4ODAwMCw3NDAwMDAsODUzMDAwLDYyNzAwMCw1NzUwMDAsMTkxMDAwLDUwNDAwMCwyNDAwMDAsNjgwMDAwLDQxMjAwMCwyOTEwMDAsNjc5MDAwLDI2OTAwMCwyNDUwMDAsNjc5MDAwLDIwNjAwMCwxMDU3MDAwLDY3OTAwMCw2NjEwMDAsNjc5MDAwLDEzNzAwMCw1MzEwMDAsMTgwMDAwLDEwMDAsNjc5MDAwLDQ5NjAwMCw1MDgwMDAsNzA5MDAwLDY3OTAwMCw2NzkwMDAsNjc5MDAwLDY1NzAwMCw3MTgwMDAsMjU2MDAwLDUwMDAsNDk4MDAwLDU1NDAwMCwxNTUwMDAsMzc4MDAwLDUwMzAwMCwzMDgwMDAsMTcwMDAsMTA1MDAwLDI1OTAwMCw3NDQwMDAsNjc5MDAwLDEwMDAsMzE2MDAwLDY3OTAwMCw2NzkwMDAsMTA3MDAwLDYwOTAwMCw2NDEwMDAsNTQwMDAwLDg5MTAwMCw2NzkwMDAsNDA4MDAwLDY3OTAwMCwzMzMwMDAsNjYwMDAwLDEyOTAwMCwxMjkwMDAsNjc5MDAwLDkwNzAwMCw4NzcwMDAsNjc5MDAwLDQxOTAwMCw1MDYwMDBdLCJvcHMiOjQ4ODMzMH1dLCJ1cGRhdGVkIjoiMjAyMy0xMi0yNVQwODozOTozMC44MDFaIn0%3D)